lib/Federation/CloudFederationProviderTalk.php

```diff
@@ -0,0 +1,263 @@
+
+	/** @var IUserManager */
+	private $userManager;
+
+	/** @var AddressHandler */
+	private $addressHandler;
+
+	/** @var FederationManager */
+	private $federationManager;
+
+	/** @var INotificationManager */
+	private $notificationManager;
+
+	/** @var IURLGenerator */
+	private $urlGenerator;
+
+	/** @var ParticipantService  */
+	private $participantService;
+
+	/** @var AttendeeMapper  */
+	private $attendeeMapper;
+
+	/** @var Manager  */
+	private $manager;
+
+	public function __construct(
+		IUserManager $userManager,
+		AddressHandler $addressHandler,
+		FederationManager $federationManager,
+		INotificationManager $notificationManager,
+		IURLGenerator $urlGenerator,
+		ParticipantService $participantService,
+		AttendeeMapper $attendeeMapper,
+		Manager $manager
+	) {
+		$this->userManager = $userManager;
+		$this->addressHandler = $addressHandler;
+		$this->federationManager = $federationManager;
+		$this->notificationManager = $notificationManager;
+		$this->urlGenerator = $urlGenerator;
+		$this->participantService = $participantService;
+		$this->attendeeMapper = $attendeeMapper;
+		$this->manager = $manager;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getShareType(): string {
+		return 'talk-room';
+	}
+
+	/**
+	 * @inheritDoc
+	 * @throws HintException
+	 * @throws DBException
+	 */
+	public function shareReceived(ICloudFederationShare $share): string {
+		if (!$this->federationManager->isEnabled()) {
+			throw new ProviderCouldNotAddShareException('Server does not support talk federation', '', Http::STATUS_SERVICE_UNAVAILABLE);
+		}
+		if ($share->getShareType() !== 'user') {
+			throw new ProviderCouldNotAddShareException('support for sharing with non-users not implemented yet', '', Http::STATUS_NOT_IMPLEMENTED);
+			// TODO: Implement group shares
+		}
+
+		if (!is_numeric($share->getShareType())) {
+			throw new ProviderCouldNotAddShareException('RoomType is not a number', '', Http::STATUS_BAD_REQUEST);
+		}
+
+		$shareSecret = $share->getShareSecret();
+		$shareWith = $share->getShareWith();
+		$roomToken = $share->getProviderId();
+		$roomName = $share->getResourceName();
+		$roomType = (int) $share->getShareType();
+		$sharedBy = $share->getSharedByDisplayName();
+		$sharedByFederatedId = $share->getSharedBy();
+		$owner = $share->getOwnerDisplayName();
+		$ownerFederatedId = $share->getOwner();
+		[, $remote] = $this->addressHandler->splitUserRemote($ownerFederatedId);
+
+		// if no explicit information about the person who created the share was send
+		// we assume that the share comes from the owner
+		if ($sharedByFederatedId === null) {
+			$sharedBy = $owner;
+			$sharedByFederatedId = $ownerFederatedId;
+		}
+
+		if ($remote && $shareSecret && $shareWith && $roomToken && $roomName && $owner) {
+			$shareWith = $this->userManager->get($shareWith);
+			if ($shareWith === null) {
+				throw new ProviderCouldNotAddShareException('User does not exist', '',Http::STATUS_BAD_REQUEST);
```

Comment:
Should bruteforce throttle here to prevent user enumeration?

Reply:
It does have brute force protection.
https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/apps/cloud_federation_api/lib/Controller/RequestHandlerController.php#L105-L107

That part is also pretty much replicating the file sharing user check
https://github.com/nextcloud/server/blob/b78f3a57d1d05fb5f81ec8361139b64a3e54d7b0/apps/federatedfilesharing/lib/OCM/CloudFederationProviderFiles.php#L230